source/language/openpulse.rst

```diff
@@ -0,0 +1,740 @@
+
+This grammar is motivated by the original OpenPulse specification, a JSON wire-format for
+pulse-level quantum programs defined in the paper
+`Qiskit Backend Specifications for OpenQASM and OpenPulse Experiment <https://arxiv.org/abs/1809.03452>`_.
+The textual format described here has several advantages over the original JSON format:
+
+- Improved readability
+- Pulse timing is based on instruction ordering and works with programs containing branching control flow
+- Reusable gate calibrations enable more succinct calibration descriptions.
+- Pulse definitions are declared as a calibration for individual circuit instructions attached to physical qubits enabling the microcoding of gate level operations.
+- Richer ability to compose complex pulses through natural DSP-like operations.
+- Clearly defined relationship between pulses, channels and the phases of compilation to hardware resources.
+- Use of multiple oscillators on a single channel at the same time (see :ref:`Geo Gate`):
+
+Openpulse provides a flexible programming model that should extend to many quantum control schemes and hardware.
+At the core of the OpenPulse grammar are the concepts of ``signals`` and ``channels``. Many of these concepts
+are inspired by the signal processing language <Faust `https://faust.grame.fr/`>_.
+``Signals`` define a discrete set of samples (or waveform) which are then played or captured via a
+a ``channel`` on the target device. A channel is a software abstraction which
+allows the programmer to be agnostic to complexities of the device's underlying pulse generation
+hardware. It is the responsibility of the target device's compiler to map ``signals``
+to the applied channels in target hardware.
+
+**NOTE: We assume we have arrays defined with a [...] symtax and complex[size] types.**
+
+Signals
+-------
+
+A ``signal`` is a generalized concept of a ``pulse``. A signal is a discrete, time-dependent function
+:math:`s(t): ℤ->ℂ`. The input time is an integer representing the number of elapsed samples. It can
+be viewed as the "index" into the signal. In hardware, the time :math:`t_h`, is given by
+:math:`t_h=t*dt`, where ``dt`` is the time elapsed per sample. Signals have a max unit-norm
+(signal processes that emit a signal outside of this range clip).
+
+All signals have a ``.duration`` property giving the length of the signal.
+
+.. code-block:: c
+
+    signal sig = [0.1, 1.+1I, 0.1 - 0.2I];
+    int[32] x = sig.duration + 1;  // x is 4
+
+Signals are produced from two fundamental types, ``envelopes`` and ``carriers``.
+
+Envelopes
+~~~~~~~~~
+
+Envelopes are discrete waveforms :math:`f(t), ℤ->ℂ`. Examples include a waveform described as
+an array of complex samples or a parametric function supported directly by the hardware
+such as a Gaussian waveform.
+
+Arrays initialization is done as in OpenQASM.
+
+.. code-block:: c
+
+    envelope env = [0.1, 1.+1I, 0.1 - 0.2I, ...];
```

Comment:
is this the same as `signal sig = `?

Reply:
Closing--will address as part of the frame proposal.